### ISSUE TITLE
[core] Possible memory leak in raylet when a required object takes a long time to fetch

### ISSUE THREAD

**Issue by stephanie-wang** [ray-project/ray#14458]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* v2.0dev

When an object is needed on a node, the raylet periodically pings its owner to check that the owner is still alive. This is a long-polling RPC that can potentially stay in the raylet's outstanding requests for a long time. This may be related to #14322, but I need to do some investigation to confirm.

**Comment by rkooo567** [comment]
Let me try making a reproduction quickly. I assume just calling ray.put million times in a driver should be able to reproduce the issue (if I understand your statement correctly)

**Comment by rkooo567** [comment]
```python3
import ray
ray.init(address='auto')
import numpy as np
refs = []
arr = np.ones(1024, dtype=np.uint8)
for _ in range(1000 * 1000):
    refs.append(ray.put(arr))

print("done")
while True:
    import time
    time.sleep(1)
```

Okay. I verified this also increases the memory usage by more than 10GB for both driver and raylet. One interesting discovery is that raylet memory usage has remained > 10GB even after the driver exits. (But when I run the driver twice, the memory usage of the raylet didn't increase... interesting).

**Comment by rkooo567** [comment]
Hmm actually, I don't see any of this log from the driver; 
```
  RAY_LOG(DEBUG) << "Received GetObjectStatus " << object_id;
```

**Comment by rkooo567** [comment]
Actually, I did one experiment with the reproduction above and found out the memory growth of the driver is pretty consistent. 

Each object 1 KB
-----------------
10 owned objects -> 67MB
100 owned objects -> 69MB (2MB+)
1000 owned objects -> 85MB (16MB+)
10,000 owned objects -> 225MB (140MB+)
100,000 owned objects -> 1.6GB (1400MB+)
1,000,000 owned objects -> 15.8 GB (14000MB+)

This indicates storing each reference consistently costs around 15KB.

**Comment by stephanie-wang** [comment]
Closed by #14184.
